File: kotlin-typescript/karakum/plugins/convertJSDocAugmentsTagClassReference.js
import ts from "typescript";
import * as karakum from "karakum";

export default function (node, context, render) {
    if (
        ts.isIndexedAccessTypeNode(node)

        && ts.isTypeReferenceNode(node.objectType)
        && ts.isIdentifier(node.objectType.typeName)
        && node.objectType.typeName.text === "JSDocAugmentsTag"

        && ts.isLiteralTypeNode(node.indexType)
        && ts.isStringLiteral(node.indexType.literal)
        && node.indexType.literal.text === "class"
    ) {
        return `JSDocAugmentsTagClass`
    }

    if (
        ts.isIndexedAccessTypeNode(node)

        && ts.isTypeReferenceNode(node.objectType)
        && ts.isIdentifier(node.objectType.typeName)
        && node.objectType.typeName.text === "JSDocImplementsTag"

        && ts.isLiteralTypeNode(node.indexType)
        && ts.isStringLiteral(node.indexType.literal)
        && node.indexType.literal.text === "class"
    ) {
        return `JSDocImplementsTagClass`
    }

    return null
}

File: kotlin-typescript/karakum/src/jsMain/kotlin/typescript/karakum/plugins/convertJSDocAugmentsTagClassReference.kt
package typescript.karakum.plugins

import arrow.core.raise.nullable
import io.github.sgrishchenko.karakum.extension.createPlugin
import typescript.isIdentifier
import typescript.isIndexedAccessTypeNode
import typescript.isLiteralTypeNode
import typescript.isStringLiteral
import typescript.isTypeReferenceNode

val convertJSDocAugmentsTagClassReference = createPlugin { node, _, _ ->
    nullable {
        ensure(isIndexedAccessTypeNode(node))

        val objectType = node.objectType
        ensure(isTypeReferenceNode(objectType))

        val objectTypeName = objectType.typeName
        ensure(isIdentifier(objectTypeName))
        ensure(objectTypeName.text == "JSDocAugmentsTag")

        val indexType = node.indexType
        ensure(isLiteralTypeNode(indexType))

        val indexTypeLiteral = indexType.literal
        ensure(isStringLiteral(indexTypeLiteral))
        ensure(indexTypeLiteral.text == "class")

        "JSDocAugmentsTagClass"
    } ?: nullable {
        ensure(isIndexedAccessTypeNode(node))

        val objectType = node.objectType
        ensure(isTypeReferenceNode(objectType))

        val objectTypeName = objectType.typeName
        ensure(isIdentifier(objectTypeName))
        ensure(objectTypeName.text == "JSDocImplementsTag")

        val indexType = node.indexType
        ensure(isLiteralTypeNode(indexType))

        val indexTypeLiteral = indexType.literal
        ensure(isStringLiteral(indexTypeLiteral))
        ensure(indexTypeLiteral.text == "class")

        "JSDocImplementsTagClass"
    }
}
